### ISSUE TITLE
Add typescript definitions

### ISSUE THREAD

**Issue by benbowes** [benbowes/react-responsive-select#75]
Improve readability of code with typescript

**Comment by robdmoore** [comment]
Would also be great to have a `@types/react-response-select` package for use of this with TypeScript

**Comment by robdmoore** [comment]
I just whipped this together and it seems to work:

```typescript
import * as React from 'react'

export interface ReactResponsiveSelectProps {
  name: string
  options: ReactResponsiveSelectOption[]
  onSubmit?: Function
  onChange?: (changes: OnChangeArgs) => void
  caretIcon?: JSX.Element
  selectedValue?: string
  prefix?: string
  disabled?: boolean
  customLabelRenderer?: (selected: ReactResponsiveSelectOption) => JSX.Element
  multiselect?: boolean
  selectedValues?: string[]
}

export interface OnChangeArgs {
  altered: boolean
  options: ReactResponsiveSelectOption[]
}

export interface ReactResponsiveSelectOption {
  text: string
  value: string
  markup?: JSX.Element
}

export default class ReactResponsiveSelect extends React.Component<ReactResponsiveSelectProps> {}


```

**Comment by robdmoore** [comment]
I didn't flesh out the `onSubmit` signature though since I'm not using that

**Comment by benbowes** [comment]
Looks good @robdmoore. With typeScript, Is generally preferable to have the types in the org `@types/*` over having them importable in the package?

**Comment by robdmoore** [comment]
Unsure. What I will say is that lately all packages I've been adding typically have the `@types/*` package and if they don't then I assume they don't have types and I think twice about using them. In this case I persisted because I had been struggling to come up with a mobile friendly filtering interface for the DDD Perth voting experience (https://github.com/dddwa/dddperth-website, https://dddperth.com/vote). I'd already been through two other libraries that I wasn't happy with and stumbled across yours and it was perfect for my needs (so thanks so much for the library - it's awesome!).

Interestingly, moment.js have the type definition as part of their package (https://github.com/moment/moment/blob/develop/moment.d.ts), but they also have a `@types/moment` package (that is empty).

**Comment by robdmoore** [comment]
If you're curious here is where I used your library: https://github.com/dddwa/dddperth-website/blob/master/components/voting.tsx#L320

**Comment by benbowes** [comment]
I guess moment have an intention to place it under `@types/*`. It certainly is easier to add it under ones own repo though 🤔

**Comment by benbowes** [comment]
@robdmoore I've made a start on this. I will add it to the project first.

**Comment by benbowes** [comment]
Though not finished, something like this...

```
import * as React from "react";

export default ReactResponsiveSelect.ReactResponsiveSelect;

declare namespace ReactResponsiveSelect {
  interface ReactResponsiveSelectProps {
    name: string;
    options: ReactResponsiveSelectOption[];
    onSubmit?: Function;
    onChange?: (changes: OnChangeArgs) => void;
    caretIcon?: JSX.Element;
    selectedValue?: string;
    prefix?: string;
    disabled?: boolean;
    customLabelRenderer?: (selected: ReactResponsiveSelectOption) => JSX.Element;
    multiselect?: boolean;
    selectedValues?: string[];
  }

  interface OnChangeArgs {
    altered: boolean;
    text: string;
    name: string;
    value: string;
    options: ReactResponsiveSelectOption[];
  }

  interface ReactResponsiveSelectOption {
    text: string;
    value: string;
    markup?: JSX.Element;
    disabled?: boolean;
  }

  export class ReactResponsiveSelect extends React.Component<ReactResponsiveSelectProps> {}
}

```

**Comment by robdmoore** [comment]
Nice!

--

**Comment by benbowes** [comment]
Released 4.2.0 to add types locally. I've updated the example at https://github.com/benbowes/react-responsive-select-with-typescript
